lib/base-compiler.js

```diff
@@ -695,6 +712,88 @@ export class BaseCompiler {
             await this.runCompiler(this.compiler.exe, newOptions, this.filename(inputFilename), execOptions));
     }
 
+    async generatePP(inputFilename, compilerOptions, rawPpOptions) {
+        // -E to dump preprocessor output, remove -o so it is dumped to stdout
+        compilerOptions = compilerOptions.concat(['-E']);
+        if (compilerOptions.includes('-o')) {
+            compilerOptions.splice(compilerOptions.indexOf('-o'), 2);
+        }
+
+        let ppOptions = _.extend({
+            'filter-headers': false,
+            'clang-format': false,
+        }, rawPpOptions);
+        
+        const execOptions = this.getDefaultExecOptions();
+        // A higher max output is needed for when the user includes headers
+        execOptions.maxOutput = 1024 * 1024 * 1024;
+        let result = await this.runCompilerRawOutput(this.compiler.exe, compilerOptions,
+                                                     this.filename(inputFilename), execOptions);
+        let output = result.stdout;
+
+        let numberOfLinesFiltered = 0;
+        if (ppOptions['filter-headers']) {
+            [numberOfLinesFiltered, output] = this.filterPP(output);
+        }
+        if (ppOptions['clang-format']) {
+            output = await this.applyClangFormat(output);
+        }
+
+        return {
+            numberOfLinesFiltered,
+            output: output,
+        };
+    }
+
+    filterPP(stdout) {
+        // Every compiler except Chibicc, as far as I've tested, outputs these line annotations
+        // Compiler test: https://godbolt.org/z/K7Pncjs4o
+        // Matching things like:
+        // # 4 "/app/example.cpp"
+        // # 11 "/usr/include/x86_64-linux-gnu/gnu/stubs.h" 2 3 4
+        // #line 1816 "C:/WinSdk/Include/10.0.18362.0/ucrt\\corecrt.h"
+        // # 13 "" 3
+        // regex test cases: https://regex101.com/r/9dOsUI/1
+        let lines = stdout.split('\n');
+        const ppLineRe = /^\s*#\s*(?:line)?\s*\d+\s*"((?:\\"|[^"])*)"/i;
+        let isInSourceRegion = true;
+        let numberOfLinesFiltered = 0;
+        let filteredLines = [];
+        for (const line of lines) {
+            let match = line.match(ppLineRe);
+            if (match === null) {
+                if (isInSourceRegion) {
+                    filteredLines.push(line);
+                } else {
+                    numberOfLinesFiltered++;
+                }
+            } else {
+                let path = match[1];
+                if (path.trim() === '' || path === '<source>' || path.endsWith('.c') || path.endsWith('.cpp')) {
```

Comment:
perhaps a silly question, but some people/projects have some very idiotic `#include "./something.cpp"` code, what kind of effect would that have, can you test so we know what to expect?

Reply:
just when i posted that I looked at twitter and seen people admitting that they've done that 🤦

Reply:
That's a good question, if a .c or .cpp file is #included its contents will not be filtered here (but any headers it #includes will be).

Reply:
that reminds me, have you tested this in IDE mode? (no cmake)

Reply:
Looks like it's working as expected:
![image](https://user-images.githubusercontent.com/51220084/153084452-cfee312c-b0ff-4eb9-a1ea-291f7571db95.png)

Reply:
bug: it says "Editor #false" in the title of the pp view

Reply:
Good eye! Looks like this affected more than just this preprocessor output pane:

![image](https://user-images.githubusercontent.com/51220084/153086451-a38f08e2-4e29-4c16-9493-83678f069f6f.png)

It appears the state object given to `Pane`'s constructor doesn't contain the tree id, only compiler name, compiler id, component name, and editor id. Since the fix looks a little more involved, perhaps I could fix this for all affected panels in one go in a separate PR?

Reply:
Understood! sounds good